validator/gulpjs/index.js

```diff
@@ -0,0 +1,149 @@
+ *
+ * @param {Object} validator - amphtml validator
+ * @returns {stream} gulp file stream
+ */
+module.exports.validate = function(validator) {
+
+  if (!validator) {
+    validator = amphtmlValidator;
+  }
+
+  function runValidation(file, encoding, callback) {
+    if (file.isNull()) {
+      return callback(null, file);
+    }
+    if (file.isStream()) {
+      this.emit('error', new PluginError(PLUGIN_NAME,
+        'Streams not supported!'));
+    }
+    if (file.isBuffer()) {
+      validator.getInstance()
+        .then(function(validatorInstance) {
+          const inputString = file.contents.toString();
+          file.ampValidationResult = validatorInstance.validateString(inputString);
+          return callback(null, file);
+        })
+        .catch(function(err) {
+          file.ampValidationResult = {
+            status: STATUS_FAIL,
+            errors: [{
+              line: 0,
+              col: 0,
+              message: err.message,
+            }],
+          };
+          return callback(null, file);
+        });
+    }
+  }
+  return through.obj(runValidation);
+};
+
+/**
+ * Formats and prints the validation results to the console.
+ *
+ * @param {Object} logger - logger used for printing the results (optional)
+ * @returns {stream} gulp file stream
+ */
+module.exports.format = function(logger) {
+
+  const results = [];
+  if (!logger) {
+    logger = gutil;
+  }
+
+  function collectResults(file, encoding, callback) {
+    if (!file.isNull() && file.ampValidationResult) {
+      results.push(file);
+    }
+    return callback(null, file);
+  }
+
+  function formatResults(callback) {
+    logger.log('AMP Validation results:\n\n' + results.map(printResult).join('\n\n'));
+    return callback();
+  }
+
+  function printResult(file) {
+    const validationResult = file.ampValidationResult;
+    let report = file.relative + ': ';
+    if (validationResult.status === STATUS_PASS) {
+      report += gutil.colors.green(validationResult.status);
+    } else {
+      report += gutil.colors.red(validationResult.status);
+    }
+    for (let ii = 0; ii < validationResult.errors.length; ii++) {
+      const error = validationResult.errors[ii];
+      let msg = file.relative + ':' + error.line + ':' + error.col + ' ' +
+        gutil.colors.red(error.message);
+      if (error.specUrl !== null) {
+        msg += ' (see ' + error.specUrl + ')';
+      }
+      report += '\n' + msg;
+    }
+    return report;
+  }
+
+  return through.obj(collectResults, formatResults);
+};
+
+/**
+ * Fail when the stream ends if any AMP validation error(s) occurred.
+ *
+ * @returns {stream} gulp file stream
+ */
+module.exports.failAfterError = function() {
+
+  let failedFiles = 0;
+
+  function collectFailedFiles(file, encoding, callback) {
+    if (file.isNull() || !file.ampValidationResult) {
+      return callback(null, file);
+    }
+    if (file.ampValidationResult.status === STATUS_FAIL) {
```

Comment:
This now makes me think that it'd actually be possible to communicate down to here that something is wrong with the validator itself, not with the file?
As in, instead of setting the status to STATUS_FAIL, could set it to some other constant that indicates validator failure. Or could set some other field, other than ampValidationResult.status. This may be something to think about because then, the failOnError thing below could say "Something wrong with the validator" instead of "Validation failed for this list of files". Sorry to be a bit of a pest with this distinction.